precompiles/pallet-xcm/src/lib.rs

```diff
@@ -260,21 +478,66 @@ where
 	}
 
 	// Helper function to convert and prepare each asset into a proper Location.
-	fn check_and_prepare_assets(
+	fn convert_assets(
 		assets: Vec<(Address, Convert<U256, u128>)>,
-	) -> Result<Vec<Asset>, PrecompileFailure> {
-		let mut assets_to_send: Vec<Asset> = vec![];
+	) -> Result<Vec<(Location, Convert<U256, u128>)>, PrecompileFailure> {
+		let mut assets_to_send: Vec<(Location, Convert<U256, u128>)> = vec![];
 		for asset in assets {
 			let asset_account = Runtime::AddressMapping::into_account_id(asset.0 .0);
 			let asset_location = LocationMatcher::convert(asset_account);
-			if asset_location == None {
+
+			if let Some(asset_loc) = asset_location {
+				assets_to_send.push((asset_loc, asset.1))
+			} else {
 				return Err(revert("Asset not found"));
 			}
-			assets_to_send.push(Asset {
-				id: AssetId(asset_location.unwrap_or_default()),
+		}
+		Ok(assets_to_send)
+	}
+
+	fn get_assets_to_send_and_remote_fees(
+		assets: Vec<(Location, Convert<U256, u128>)>,
+		remote_fees_id_index: Option<usize>,
+	) -> Result<(Assets, Option<AssetId>), PrecompileFailure> {
+		let assets_to_send: Assets = assets
+			.into_iter()
+			.map(|asset| Asset {
+				id: AssetId(asset.0),
 				fun: Fungibility::Fungible(asset.1.converted()),
 			})
+			.collect::<Vec<Asset>>()
+			.into();
+
+		if let Some(index) = remote_fees_id_index {
+			let remote_fees_id: AssetId = {
+				let asset = assets_to_send
+					.get(index)
+					.ok_or_else(|| RevertReason::custom("remote_fees_id not found"))?;
+				AssetId(asset.id.0.clone())
+			};
+			return Ok((assets_to_send, Some(remote_fees_id)));
+		}
+
+		Ok((assets_to_send, None))
+	}
+
+	fn check_transfer_type_is_not_reserve(
+		transfer_type: u8,
+	) -> Result<TransferType, PrecompileFailure> {
+		let transfer_type_helper: TransferTypeHelper = TransferTypeHelper::decode(
+			&mut transfer_type.to_le_bytes().as_slice(),
+		)
+		.map_err(|_| RevertReason::custom("Failed decoding value for TransferTypeHelper"))?;
+
+		match transfer_type_helper {
+			TransferTypeHelper::Teleport => return Ok(TransferType::Teleport),
+			TransferTypeHelper::LocalReserve => return Ok(TransferType::LocalReserve),
+			TransferTypeHelper::DestinationReserve => return Ok(TransferType::DestinationReserve),
+			TransferTypeHelper::RemoteReserve => {
```

Comment:
you can remove this case

Reply:
Done in aea402dd3193edd09242b0be19e07e3da4e07200.